test/test_attn.py

```diff
@@ -0,0 +1,53 @@
+import torch
+from torch.testing._internal.common_utils import (TestCase, gradcheck, gradgradcheck)
+from torch.autograd import Function
+
+class TestAutogradAttn(TestCase):
+
+    def test_autograd_attn(self):
+        class Attn(Function):
+
+            @staticmethod
+            def forward(q, k, v):
+                if q.dim() != 2:
+                    raise RuntimeError('attention function is currently supported only for 2D tensors')
+                if k.dim() != 2:
+                    raise RuntimeError('attention function is currently supported only for 2D tensors')
+                if v.dim() != 2:
+                    raise RuntimeError('attention function is currently supported only for 2D tensors')
+
+                if q.shape[-1] != k.shape[-1]:
+                    raise RuntimeError(f'Expected size of k to be {q.shape[-1]} but got: {k.shape[-1]}')
+                x = torch.matmul(q, k.transpose(1, 0))
+                a = torch.tanh(x)
+                o = torch.matmul(a, v)
+                return o, a, x
+
+            @staticmethod
+            def setup_context(ctx, inputs, outputs):
+                q, k, v = inputs
+                o, a, x = outputs
+                ctx.save_for_backward(q, k, v, a, x)
+
+            @staticmethod
+            def backward(ctx, grad_o, grad_a, grad_x):
+                q, k, v, a, x = ctx.saved_tensors
+                # TODO Figure out why we need to add `grad_a` here or `grad_x` in next step
+                # More boradly, what is `grad_a` and how gradcheck works.
```

Comment:
grad_a is the gradient wrt to the output a.
gradcheck will make sure that all gradient of every output wrt every input must be correct.
So here it looks very similar to the case where a is used in two places (to compute o and as itself). So the gradient for a should be the sum of the two contributions (backward from grad_o and grad_a itself). But this is confusing as "a" has two meanings in that sentence! one is the output of the function for which we get grad_a and one is the intermediary for which you compute dL_da.

Reply:
this helps, thanks!